### ISSUE TITLE
[Feature Request] Pass custom options [commands] through the GUI

### ISSUE THREAD

**Issue by nero-0** [firedm/FireDM#281]
Hi.
I noticed that it is possible to pass some shell commands during the post-processing step of the download, but there is no option to do this with yt-dl/yt-dlp before the extractor start working. Could you add this? I believe that adding all the options would make the GUI very polluted, so it seems a good alternative. 😃

Best regards

**Comment by firedm** [comment]
@nero-0 
not all options/flags can be passed to youtube-dl since it is used for extraction only, not for downloading nor post-processing
so blindly pass options/flags from user to youtube-dl might change its expected behavior

can you give some examples of missing options?

**Comment by nero-0** [comment]
Some that I use:
--remux-video
--audio-format
--audio-quality
--parse-metadata/--replace-in-metadata (Modify some metadata obtained by the extractor)
--embed-subs (I know FireDM can download subs, but I believe it is not possible to embed them in the container)
--cache-dir/--no-cache-dir (Sometimes FireDM leaves behind some files in C:\Users\<Username>\.cache)
--netrc/--netrc-location (In order to use .netrc authentication data)
--concurrent-fragments (Don't know if this is the same as "Connections per download" that has in FireDM)
--fragment-retries

yt-dlp also provides extra information about each stream, which makes it easier to identify in cases where multiple languages are delivered. If you can add the output of this information in the logs, or in the "Video streams" field that would be very nice. :)

Stay awesome

**Comment by candrapersada** [comment]
--merge-output-format

**Comment by firedm** [comment]
@nero-0 
thanks for reporting

> --cache-dir/--no-cache-dir (Sometimes FireDM leaves behind some files in C:\Users<Username>.cache)

we can make it  "--no-cache-dir" by default in FireDM

> --concurrent-fragments (Don't know if this is the same as "Connections per download" that has in FireDM)

yes, this is the same

> --merge-output-format

you can select the output format if you change extension of the output file name, from gui main tab ext. field, or using "-o" flag in cmdline

for `--netrc/--netrc-location` and -`-embed-subs` both can be added to FireDM settings

regarding other options, it is mostly for post processing, and can't be passed

IMO, if you need too much fine tuning, it is better to use yt-dlp as a cmdline tool in conjunction with aria2c for downloading

**Comment by candrapersada** [comment]
> you can select the output format if you change extension of the output file name, from gui main tab ext. field,

but change extension of the output can't be done in playlist.

**Comment by nero-0** [comment]
> regarding other options, it is mostly for post processing, and can't be passed

Just out of curiosity, why? I guess everything is done with ffmpeg.

> IMO, if you need too much fine tuning, it is better to use yt-dlp as a cmdline tool in conjunction with aria2c for downloading

I've tried it before, but aria2c doesn't seem to handle my connection very well.


Anyway, thanks for having your eyes on it. :)

**Comment by firedm** [comment]
> > regarding other options, it is mostly for post processing, and can't be passed
> 
> Just out of curiosity, why? I guess everything is done with ffmpeg.

you're right, these jobs is done by ffmpeg, it is the way you pass it, i mean ytdlp or youtube-dl their job finish early by extracting video info, and then get terminated, so any post processing options passed to them will be useless

to make picture clear on how firedm works, there is three main stages:

1. extracting file info, which done by youtube-dl, or yt-dlp.
2. downloading file, done by libcurl.
3. post processing, e.g. muxing, writing metadata, etc..., done by ffmpeg thru firedm logic.



> I've tried it before, but aria2c doesn't seem to handle my connection very well.

if this is the case, There might be a way out, i will do more tests if we can delegate stage-3 (post processing) to youtube-dl/yt-dlp
not sure if it will work as expected, but it worth give it a try, then we can pass any options directly to  youtube-dl/yt-dlp without a problem

will let you know if there is any progress

**Comment by nero-0** [comment]
> to make picture clear on how firedm works, there is three main stages:

Ohh, got it. Thanks for the explanation!

> if this is the case, There might be a way out, i will do more tests if we can delegate stage-3 (post processing) to youtube-dl/yt-dlp

That would be absolutely amazing! :satisfied:
Thanks a lot for your efforts.

**Comment by candrapersada** [comment]
> 3. post processing, e.g. muxing, writing metadata, etc..., done by ffmpeg thru firedm logic.


add ffmpeg latest version for FireDM in Windows.

**Comment by firedm** [comment]
> add ffmpeg latest version for firedm in windows.

@candrapersada 
thanks for reporting, will add the latest ffmpeg to new releases
